File: app/src/main/java/com/example/pupilx/managers/CalculationManager.kt
package com.example.pupilx.managers

import kotlin.math.pow
import kotlin.math.round
import kotlin.math.sqrt

class CalculationManager {

    var baseValue: String = ""
    var constrictedValue: String = ""
    var dilationValue: String = ""
    var constricted_Velocity_Value: String = ""
    var dilation_velocity_Value: String = ""
    var percentage_Change_Value: String = ""
    var Amplitude_Constriction: String = ""
    var Amplitude_Dilation: String = ""

    fun ClearAllVariables() {
        baseValue = ""
        constrictedValue = ""
        dilationValue = ""
        constricted_Velocity_Value = ""
        dilation_velocity_Value = ""
        percentage_Change_Value = ""
        Amplitude_Constriction = ""
        Amplitude_Dilation = ""
    }

    fun CalucationVelocityValue(
        baseList: List<Float>,
        constrictedList: List<Float>,
        dilationList: List<Float>,
        timeStamp_BaseList: List<Float>,
        timeStamp_ConstrictedList: List<Float>,
        timeStamp_DilationList: List<Float>
    ) {

        ClearAllVariables()

        var BaseValue = 0f
        var Base_Standard_Value = 0.0
        if (baseList.isNotEmpty()) {
            BaseValue = FindAverageValue(baseList)
            Base_Standard_Value = FindStandardDeviation(baseList)
        }

        var ConstrictedValue = 0f
        var Constricted_Standard_Value = 0.0
        if (constrictedList.isNotEmpty()) {
            Constricted_Standard_Value = FindStandardDeviation(constrictedList)
        }

        var DilationValue = 0f
        var Dilation_Standard_Value = 0.0
        if (dilationList.isNotEmpty()) {
            Dilation_Standard_Value = FindStandardDeviation(dilationList)
        }

        val temp = " +/- "

        if (baseList.isNotEmpty() && constrictedList.isNotEmpty() && dilationList.isNotEmpty()) {
            // Constriction velocity Value Start here............
            val lastBaseValue_TimeStamp = timeStamp_BaseList.last()

            val smallestConstrictedValue = constrictedList.filter { it != 0f }.minOrNull() ?: 0f
            val smallestConstrictedValue_index = constrictedList.indexOf(smallestConstrictedValue)
            val smallestConstrictedValue_Timestamp = timeStamp_ConstrictedList[smallestConstrictedValue_index]

            val tempCon = round(smallestConstrictedValue * 1000.0f) * 0.001f
            ConstrictedValue = tempCon

            // Dilation velocity Value Start here............
            val largeDilationValue = dilationList.filter { it != 0f }.maxOrNull() ?: 0f
            val largeDilationValue_index = dilationList.indexOf(largeDilationValue)
            val largeDilationValue_Timestamp = timeStamp_DilationList[largeDilationValue_index]

            val tempDila = round(largeDilationValue * 1000.0f) * 0.001f
            DilationValue = tempDila

            // Amplitude Constriction Value
            val temp2 = (BaseValue - ConstrictedValue)
            val tempACon = round(temp2 * 1000.0f) * 0.001f
            Amplitude_Constriction = tempACon.toString()

            var temp3 = (lastBaseValue_TimeStamp - smallestConstrictedValue_Timestamp)
            temp3 /= 1000f

            var constrictedVelocity = temp2 / temp3
            val newTemp1 = round(constrictedVelocity * 1000.0f) * 0.001f
            constrictedVelocity = newTemp1

            // Amplitude Dilation Value
            val temp5 = (smallestConstrictedValue - largeDilationValue)
            val tempADila = round(temp5 * 1000.0f) * 0.001f
            Amplitude_Dilation = tempADila.toString()

            var temp4 = (smallestConstrictedValue_Timestamp - largeDilationValue_Timestamp)
            temp4 /= 1000f

            var dilationVelocity = temp5 / temp4
            dilationVelocity = round(dilationVelocity * 1000.0f) * 0.001f

            if (constrictedVelocity == 0f) {
                constrictedVelocity = 0f
            }
            if (dilationVelocity == 0f) {
                dilationVelocity = 0f
            }

            baseValue = BaseValue.toString() + temp + Base_Standard_Value.toString()
            constrictedValue = ConstrictedValue.toString() + temp + Constricted_Standard_Value.toString()
            dilationValue = DilationValue.toString() + temp + Dilation_Standard_Value.toString()
            constricted_Velocity_Value = constrictedVelocity.toString()
            dilation_velocity_Value = dilationVelocity.toString()
        }

        // new percentage Change Value............................
        val tempPercent = (BaseValue - ConstrictedValue) / BaseValue
        val newTemp = tempPercent * 100
        val tempPercen = round(newTemp * 1000.0f) * 0.001f
        percentage_Change_Value = tempPercen.toString()
    }

    fun FindAverageValue(diameters: List<Float>): Float {
        return if (diameters.isNotEmpty()) {
            val average = diameters.average().toFloat()
            round(average * 100.0f) * 0.01f
        } else {
            0.0f
        }
    }

    fun FindStandardDeviation(diameters: List<Float>): Double {
        return if (diameters.size > 1) {
            val avg = diameters.average()
            val sum = diameters.sumOf { (it - avg).pow(2) }
            round(sqrt(sum / (diameters.size - 1)) * 1000.0) / 1000.0
        } else {
            0.0
        }
    }
}

File: csharp_scripts/CalculationManager.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public class CalculationManager : MonoBehaviour
{


    public string baseValue;
    public string constrictedValue;
    public string dilationValue;
    public string constricted_Velocity_Value;
    public string dilation_velocity_Value;
    public string percentage_Change_Value;
    public string Amplitude_Constriction;
    public string Amplitude_Dilation;


    //  float SmallestConstrictedValue;
    //  float largeDilationValue;
    //  float ConstrictedVelocity;
    //  float DilationVelocity;

    // Start is called before the first frame update
    void Start()
    {

    }

    public void ClearAllVariables()
    {
        baseValue = "";
        constrictedValue = "";
        dilationValue = "";
        constricted_Velocity_Value = "";
        dilation_velocity_Value = "";

        percentage_Change_Value = "";
        Amplitude_Constriction = "";
        Amplitude_Dilation = "";
    }



    public void CalucationVelocityValue(List<float> baseList, List<float> constrictedList, List<float> dilationList, List<float> timeStamp_BaseList, List<float> timeStamp_ConstrictedList, List<float> timeStamp_DilationList)
    {

        // first clear all variables here 
        ClearAllVariables();


        float BaseValue = 0;
        double Base_Standard_Value = 0;
        if (baseList.Count > 0)
        {
            BaseValue = FindAverageValue(baseList);
            Base_Standard_Value = FindStandardDeviation(baseList);
        }

        float ConstrictedValue = 0;
        double Constricted_Standard_Value = 0;
        if (constrictedList.Count > 0)
        {
            // ConstrictedValue = FindAverageValue(constrictedList);
            Constricted_Standard_Value = FindStandardDeviation(constrictedList);
        }

        float DilationValue = 0;
        double Dilation_Standard_Value = 0;
        if (dilationList.Count > 0)
        {
            // DilationValue = FindAverageValue(dilationList);
            Dilation_Standard_Value = FindStandardDeviation(dilationList);
        }


        string temp = " +/- ";

        if (baseList.Count > 0 && constrictedList.Count > 0 && dilationList.Count > 0)
        {
            //Constrion velocity Value Start here............
            float lastBaseValue_TimeStamp = timeStamp_BaseList[timeStamp_BaseList.Count - 1];

            float SmallestConstrictedValue = constrictedList.Where(x => x != 0).DefaultIfEmpty().Min();
            int SmallestConstrictedValue_index = constrictedList.IndexOf(SmallestConstrictedValue);
            float SmallestConstrictedValue_Timestamp = timeStamp_ConstrictedList[SmallestConstrictedValue_index];


            float Temp_con = Mathf.Round(SmallestConstrictedValue * 1000.0f) * 0.001f;
            ConstrictedValue = Temp_con;


            //Dilation velocity Value Start here............
            float largeDilationValue = dilationList.Where(x => x != 0).DefaultIfEmpty().Max();
            int largeDilationValue_index = dilationList.IndexOf(largeDilationValue);
            float largeDilationValue_Timestamp = timeStamp_DilationList[largeDilationValue_index];

            float Temp_dila = Mathf.Round(largeDilationValue * 1000.0f) * 0.001f;
            DilationValue = Temp_dila;

            //Amplitude Constriction Value
            float temp2 = (BaseValue - ConstrictedValue);

            float Temp_A_Con = Mathf.Round(temp2 * 1000.0f) * 0.001f;
            Amplitude_Constriction = Temp_A_Con.ToString();

            float temp3 = (lastBaseValue_TimeStamp - SmallestConstrictedValue_Timestamp);

            temp3 = temp3 / 1000;

            float ConstrictedVelocity = temp2 / temp3;

            float newTemp1 = Mathf.Round(ConstrictedVelocity * 1000.0f) * 0.001f;

             ConstrictedVelocity = newTemp1;


            //Amplitude Dialation Value
            float temp5 = (SmallestConstrictedValue - largeDilationValue);

            float Temp_A_Dila = Mathf.Round(temp5 * 1000.0f) * 0.001f;
            Amplitude_Dilation = Temp_A_Dila.ToString();

            float temp4 = (SmallestConstrictedValue_Timestamp - largeDilationValue_Timestamp);

            temp4 = temp4 / 1000;

            float DilationVelocity = temp5 / temp4;

               DilationVelocity = Mathf.Round(DilationVelocity * 1000.0f) * 0.001f;


            if (ConstrictedVelocity != 0)
            {

            }
            else
            {
                ConstrictedVelocity = 0;
            }
            if (DilationVelocity != 0)
            {

            }
            else
            {
                DilationVelocity = 0;
            }

            baseValue = BaseValue + temp + Base_Standard_Value;
            constrictedValue = ConstrictedValue + temp + Constricted_Standard_Value;
            dilationValue = DilationValue + temp + Dilation_Standard_Value;
            constricted_Velocity_Value = "" + ConstrictedVelocity;
            dilation_velocity_Value = "" + DilationVelocity;

        }

        // neww percentage Change Value............................

        float Temp = (BaseValue - ConstrictedValue) / BaseValue;

        float newTemp = Temp * 100;

        float temp_percen =  Mathf.Round(newTemp * 1000.0f) * 0.001f;

        percentage_Change_Value = "" + temp_percen;

    }


    public float FindAverageValue(List<float> diameters)
    {
        // float average = diameters.Count > 0 ? diameters.Average() : 0.0;
        float average;

        if (diameters.Count > 0)
        {
            average = diameters.Average();
            average = Mathf.Round(average * 100.0f) * 0.01f;
        }
        else
        {
            average = 0.0f;
        }
        //  Debug.Log("FindAverageValue :" + average);

        return average;
    }
    public double FindStandardDeviation(List<float> diameters)
    {
        double average;
        if (diameters.Count > 0)
        {
            double avg = diameters.Average();

            // Perform the Sum of (value-avg)_2_2.      
            double sum = diameters.Sum(d => Math.Pow(d - avg, 2));

            // Put it all together.      
            average = Math.Sqrt((sum) / (diameters.Count() - 1));

            average = Math.Round(average, 3, MidpointRounding.AwayFromZero);
            //  Debug.Log("avarage stand =" + average);
        }
        else
        {
            average = 0.0f;
        }

        //   Debug.Log("FindStandardDeviation_v2 :" + average);
        return average;


    }
}

